### DISCUSSION TITLE
Capture from terminal

### DISCUSSION THREAD

**Discussion by jgollenz** [nvim-orgmode/orgmode#230] [Q&A]
This may be considered golfing. I often find myself outside of vim when I want to capture something in orgmode. Being naturally ~~lazy~~ efficient I looked at my workflow

1. open vim (`v`)
2. shortcut to open capture buffer (`<leader>oc`)
3. capture
4. close capture (`:wq`)
5. close vim (`:q`)

and it's less than desirable. Yes I can collapse steps 4 and 5 to `:wqa` and I will relearn my fingers to do so (perhaps you know an even better way than `:wqa`?). But still, I already know that I want to only make a capture and nothing else before step 1. So I researched if I could do something like `v -"<leader>oc"` to jump directly into the capture buffer. Indeed, there is `-c|--cmd` that allows me to do this `v +"norm 1 oc"` (see [this question](https://vi.stackexchange.com/questions/7779/how-to-use-leader-in-a-normal-command) for why the `1` is necessary when your `<leader>` is mapped to space). But alas! It does not work for this particular remapping. The `norm 1 ...` syntax works for other mappings I have and also for e.g. `norm 1 oih`. I have made no research yet why it does not work for for `norm 1 oc`, but it also doesn't work for `norm 1 oa` (opening the agenda), so I suspect it's something about opening a buffer. 

The reason I write here is that perhaps I'm doing this all wrong and there is a really easy way to start the capture buffer straight from the terminal.

**Comment by kristijanhusak** [comment]
You can probably write your own wrapper around these calls to make your life easier. Something like this:

```lua
-- Orgmode setup call. This is needed in the function
local org = require('orgmode').setup(require('partials.orgmode_config'))

function _G.capture_from_terminal()
  -- Ensure org is initialized
  org:init()

  vim.schedule(function()
    -- Open desired template
    org.capture:open_template_by_shortcut('t')
    -- Map Q to write + exit Neovim. Make sure to leave vim.schedule calls
    -- to make sure everything is written
    vim.keymap.set('n', 'Q', function()
      vim.cmd[[wq]]
      vim.schedule(function()
        vim.cmd[[qall!]]
      end)
    end)
  end)
end

```

And then run your nvim with `nvim -c "lua capture_from_terminal()"`, write your notes, and press `Q` in normal mode.

Note that this touches some of the internal things, so I'm not guaranteeing this will not break in future.
But I believe it should work for the most part.

**Reply by jgollenz** [reply]
Thanks a lot! Been tinkering around with it but it keeps crashing on `vim.keymap`, telling me that it is null. And when I comment out the mapping part, there does not seem to be a template with shortcut `t` (or any other shortcut). Almost like something might not be loaded/initialized? Any idea what could cause that? My nvim version is 0.6.0. What works fine is just doing `org.capture:prompt()` (shows all my templates+shortcuts correctly) instead, which what I was envisioning originally. But your approach with specifying the template is even better since I can wrap it in a bash function that I would call like so `c <template_shortcut>`

The only thing I have different from your snippet is that my config is in the same file, `.setup({...})`

**Reply by jgollenz** [reply]
Hm, turns out that `c<CR>t` is faster than `c<SPACE>t<CR>`. I will still try to find out what the issue is about, but I probably just stick to invoking `org.capture:prompt()` :)

**Reply by kristijanhusak** [reply]
`vim.keymap` is available only on 0.7.0-dev. If you are using stable version, you need to replace it with `nvim_buf_set_keymap`.
Regarding the shortcut, do you have option `t` when you open up the prompt? I used that since it's built in for capturing to refile file, but maybe you override it.

How you do setup doesn't matter, I just put mine as an example.

**Reply by jgollenz** [reply]
I see, that explains that. Will try with 0.7.0 when I get the time. 

> do you have option t when you open up the prompt

I left that at the default, so `t` should be there. But I tried also with other shortcuts for my custom templates. Didn't work either.

Fixed it with #231

**Reply by kristijanhusak** [reply]
Merged it, should be working now.

**Comment by lanker** [comment]
I use `nvim +"lua require('orgmode').action('capture.prompt')"` to open nvim in capture mode.

**Reply by jgollenz** [reply]
That's very clean, I like it :) made an alias for `nvim +"lua require('orgmode').action('agenda.agenda')"` as well

Thank you!
